### ISSUE TITLE
allow selecting text in normal mode without switching over to copy-mode

### ISSUE THREAD

**Issue by timotheecour** [tmux/tmux#732]
This:
`bind -t vi-copy MouseDragEnd1Pane copy-selection -x`
allows not loosing scrollback position when selecting text in copy mode, but has side effect of entering copy mode when selecting text in normal mode, and it breaks the workflow; could we have an option to allow selecting text in normal mode without switching over to copy-mode ?

For context, see here: https://github.com/tmux/tmux/issues/337#issuecomment-223753803
> I agree it's slightly surprising selecting text in normal mode switches over to copy-mode, instead of simply filling the copy buffer with the selection and staying in normal mode. The extra q or Enter keypress to exit from the unexpected copy-mode breaks the workflow. But it's already quite an improvement, not being bounced back while in highlighting during copy-mode!

**Comment by nicm** [comment]
You should be able to script this, you can set a user option when you enter copy mode from the -n binding, and then check it in the MouseDragEnd1Pane binding.

**Comment by timotheecour** [comment]
thanks @nicm ; do you have any pointers that would help ? not super familiar with tmux's scripting syntax

**Comment by nicm** [comment]
What bindings are you already using? By default if you just change the MouseDrag1Pane to enter copy-mode then it will exit when the selection finishes.

**Comment by timotheecour** [comment]
@nicm 
> What bindings are you already using?
```
bind -t vi-copy MouseDragEnd1Pane copy-selection -x
```
(cf 1st post); or are you asking something else?

btw, this was what you had recommended here: https://github.com/tmux/tmux/issues/337#issuecomment-198670586

> By default if you just change the MouseDrag1Pane to enter copy-mode then it will exit when the selection finishes.

Not sure what you're suggesting, a full command would really help (me as well as others, eg https://github.com/tmux/tmux/issues/337#issuecomment-223753803)

I also tried this, which allows selecting text in normal mode and returning to normal mode once the mouse is released, but is even worse because of https://github.com/tmux/tmux/issues/337#issuecomment-198670586 (exits copy mode after selecting text in copy mode and releasing mouse):
```
bind-key -t vi-copy MouseDragEnd1Pane copy-pipe "xclip -in -selection clipboard"
```

To summarize: what are the commands to put in tmux.conf that allow:
* when in normal mode, selecting text goes back to normal mode upon mouse release
* when in copy mode, selecting text stays in copy mode upon mouse release

(in other words, selecting text and mouse release wouldn't alter the state (normal/copy) the user was in)

Thanks for your help

**Comment by nicm** [comment]
You can do something like (using master not 2.3, use copy-mode-vi rather than copy-mode for vi keys as usual):

~~~~
bind -Tcopy-mode MouseDrag1Pane setw @mode 1\; copy-mode -M
bind -Tcopy-mode MouseDragEnd1Pane if -F '#{@mode}' 'send -X copy-selection' 'send -X copy-selection-and-cancel'
bind-key -n MouseDrag1Pane if -F '#{mouse_any_flag}' 'if -F "#{pane_in_mode}" "setw -u @mode; copy-mode -M" "send-keys -M"' "setw -u @mode; copy-mode -M"
~~~~

Which works so long as you only go into copy mode in one pane in the window at a time. That would be fixable, but you would need to write a shell script because there is currently no way to generate per-pane user option names from inside tmux.

**Comment by nicm** [comment]
In fact it should work even with multiple panes in copy mode, so long as you don't drag a selection with the mouse in two at once, which would be impossible anyway :-).

**Comment by nicm** [comment]
Actually, no, a drag in a different pane that is not in copy mode will clear the flag that existing copy modes will later check. Anyway, if that is an issue you will need a script or some other magic to save/restore the flag.

**Comment by timotheecour** [comment]
@nicm could you please provided updated instructions for tmux 2.5?
i tried this (and many variants) on OSX / vi and doesn't work:

```
bind-key -T copy-mode-vi MouseDrag1Pane setw @mode 1\; copy-mode -M
bind-key -T copy-mode-vi MouseDragEnd1Pane if -F '#{@mode}' 'send-keys -X copy-pipe "reattach-to-user-namespace pbcopy"' 'send-keys -X copy-selection-and-cancel "reattach-to-user-namespace pbcopy"'
bind-key -n MouseDrag1Pane if -F '#{mouse_any_flag}' 'if -F "#{pane_in_mode}" "setw -u @mode; copy-mode -M" "send-keys -M"' "setw -u @mode; copy-mode -M"

```

**Comment by ndeubert** [comment]
These commands don't seem to work on 2.7 either. Is there a new way to do this?
